pkg/kubelet/kubelet_node_status.go

```diff
@@ -53,36 +52,35 @@ const (
 	maxNamesPerImageInNodeStatus = 5
 )
 
-// registerWithApiServer registers the node with the cluster master. It is safe
+// registerWithApiServerWithRetry registers the node with the cluster master. It is safe
 // to call multiple times, but not concurrently (kl.registrationCompleted is
 // not locked).
-func (kl *Kubelet) registerWithApiServer() {
+func (kl *Kubelet) registerWithApiServerWithRetry() {
 	if kl.registrationCompleted {
 		return
 	}
-	step := 100 * time.Millisecond
 
-	for {
-		time.Sleep(step)
-		step = step * 2
-		if step >= 7*time.Second {
-			step = 7 * time.Second
-		}
-
-		node, err := kl.initialNode()
-		if err != nil {
-			glog.Errorf("Unable to construct v1.Node object for kubelet: %v", err)
-			continue
-		}
+	var completedNode *v1.Node = nil
 
-		glog.Infof("Attempting to register node %s", node.Name)
-		registered := kl.tryRegisterWithApiServer(node)
-		if registered {
-			glog.Infof("Successfully registered node %s", node.Name)
-			kl.registrationCompleted = true
-			return
-		}
-	}
+	registerInitiallyWithInfiniteRetry(kl.clock,
+		func() bool {
+			// Possible error can happen when querying cloud info, so may need retrying.
+			node, err := kl.tryPopulateInitialNodeData()
+			if err != nil {
+				glog.Errorf("Unable to construct v1.Node object for kubelet: %v", err)
+				return false
+			} else {
+				completedNode = node
+				return true
+			}
+		}, "populate initial node with metadata")
+
+	registerInitiallyWithInfiniteRetry(kl.clock,
```

Comment:
Because then we are unnecessarily re-registering or else we need to couple metadata acquisition logic to the state of the registration.